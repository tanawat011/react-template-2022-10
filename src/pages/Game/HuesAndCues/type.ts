/* istanbul ignore file */
import type { DocumentData, QueryDocumentSnapshot } from 'firebase/firestore'

export type Player = {
  id: string
  name: string
}

export type Room = {
  id: string
  password: string
  hintChoice: string[]
  hintSelected: string
  isStarted: boolean
  isSubmitResult: boolean
  totalRound: number
}

export type RawRoomPlayer = {
  refPlayer: QueryDocumentSnapshot<DocumentData>
  color: string
  score: number
  allSelected: string[]
  isOwner: boolean
  isTurn: boolean
  isHinter: boolean
  totalTurn: number
  seq: number
}

export type RoomPlayer = Omit<RawRoomPlayer, 'refPlayer'> & {
  player: Player
}
