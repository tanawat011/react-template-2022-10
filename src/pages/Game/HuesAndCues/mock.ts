/* istanbul ignore file */
import type { Player, Room, RoomPlayer } from './type'

type OptionAllPlayers = {
  p1?: Partial<RoomPlayer>
  p2?: Partial<RoomPlayer>
  p3?: Partial<RoomPlayer>
}

export const getMockHintChoice = ['A1', 'A2', 'A3', 'A4']

export const getMockRoom = (option?: Partial<Room>): Room => {
  return {
    id: 'room-id-x',
    password: '',
    hintChoice: [],
    hintSelected: '',
    isStarted: false,
    isSubmitResult: false,
    totalRound: 0,
    ...option,
  }
}

export const getMockPlayers = (): Player[] => {
  return [
    {
      id: 'player-id-1',
      name: 'p1',
    },
    {
      id: 'player-id-2',
      name: 'p2',
    },
    {
      id: 'player-id-3',
      name: 'p3',
    },
  ]
}

const defaultPlayerData = {
  score: 0,
  allSelected: [],
  isHinter: false,
  isOwner: false,
  isTurn: false,
  totalTurn: 0,
}

export const getMockPlayersInRoom = (option?: OptionAllPlayers): RoomPlayer[] => {
  const players = getMockPlayers()

  return [
    {
      player: players[0],
      color: 'bg-[#000000]',
      seq: 1,
      ...defaultPlayerData,
      ...option?.p1,
    },
    {
      player: players[1],
      color: 'bg-[#999999]',
      seq: 2,
      ...defaultPlayerData,
      ...option?.p2,
    },
    {
      player: players[2],
      color: 'bg-[#666666]',
      seq: 3,
      ...defaultPlayerData,
      ...option?.p3,
    },
  ]
}
