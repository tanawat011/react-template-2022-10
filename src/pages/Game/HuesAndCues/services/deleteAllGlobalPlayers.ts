import type { RoomPlayer } from '../type'

import { FIRESTORE_PATH } from 'constants/gameHuesAndCues'
import { deleteDocument } from 'helpers/firebase'

export const deleteAllGlobalPlayers = async (players: RoomPlayer[]) => {
  const playerPath = FIRESTORE_PATH.DOC_PLAYER

  await Promise.all(
    players.map(async ({ player: { id } }) => {
      const path = playerPath + `/${id}`

      await deleteDocument(path)
    }),
  )
}
