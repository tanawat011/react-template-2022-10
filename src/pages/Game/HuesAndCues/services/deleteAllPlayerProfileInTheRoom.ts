import type { RoomPlayer } from '../_type'

import { getFirestore } from 'firebase/firestore'

import { deleteDocument } from 'helpers/firebase'

import { FIRESTORE_PATH } from '../_constants'

export const deleteAllPlayerProfileInTheRoom = async (players: RoomPlayer[]) => {
  const db = getFirestore()

  await Promise.all(
    players.map(async ({ uid }) => {
      const collectionPath = `${FIRESTORE_PATH.PLAYERS_COLLECTION}/${uid}`

      await deleteDocument(db, collectionPath)
    }),
  )
}
