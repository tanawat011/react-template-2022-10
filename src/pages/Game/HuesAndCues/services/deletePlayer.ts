import { FIRESTORE_PATH } from 'constants/gameHuesAndCues'
import { deleteCollection } from 'helpers/firebase'

export const deletePlayer = async (playerId: string) => {
  const path = FIRESTORE_PATH.DOC_PLAYER + `/${playerId}`

  await deleteCollection(path)
}
