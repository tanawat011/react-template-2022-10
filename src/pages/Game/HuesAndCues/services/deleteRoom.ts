import { FIRESTORE_PATH } from 'constants/gameHuesAndCues'
import { deleteCollection } from 'helpers/firebase'

export const deleteRoom = async (roomId: string) => {
  const path = FIRESTORE_PATH.DOC_ROOM + `/${roomId}`

  await deleteCollection(path)
}
