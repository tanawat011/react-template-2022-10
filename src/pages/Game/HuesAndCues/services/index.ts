export * from './deleteAllPlayerProfileInTheRoom'
export * from './deleteRoom'
export * from './getAllPlayersInTheRoom'
export * from './getPlayerInTheRoom'
export * from './getPlayerProfile'
export * from './setPlayerProfile'
export * from './setPlayerInTheRoom'
export * from './subscribeRoom'

export * from './navigate'
