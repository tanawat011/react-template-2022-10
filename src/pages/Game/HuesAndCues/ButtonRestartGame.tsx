import type { Room, RoomPlayer } from './type'

import { Button } from 'components/Button'

import { setRoom, setRoomPlayer } from './services'

type Prop = {
  room: Room
  roomPlayers: RoomPlayer[]
  currRoomPlayer: RoomPlayer
}

// * This button will clickable when the player is a `owner`, So in this button will call the current player is a `owner`
export const ButtonRestartGame: React.FC<Prop> = ({ room, roomPlayers, currRoomPlayer }) => {
  const handleClickRestartGame = async () => {
    await setRoom({
      ...room,
      isStarted: false,
      hintChoice: [],
      hintSelected: '',
      isSubmitResult: false,
      totalRound: 0,
    })

    const owner = currRoomPlayer

    await setRoomPlayer(room.id, {
      ...owner,
      allSelected: [],
      isHinter: true,
      isTurn: true,
      score: 0,
      totalTurn: 0,
    })

    await Promise.all(
      roomPlayers
        .filter((p) => !p.isOwner)
        .map(
          async (player) =>
            await setRoomPlayer(room.id, {
              ...player,
              allSelected: [],
              isHinter: false,
              isTurn: false,
              score: 0,
              totalTurn: 0,
            }),
        ),
    )
  }

  const isOwner = currRoomPlayer.isOwner

  // * Disabled when started game or you are not a `owner`
  const isDisabled = [!isOwner].includes(true)

  return (
    <Button disabled={isDisabled} onClick={handleClickRestartGame}>
      Restart Game
    </Button>
  )
}
