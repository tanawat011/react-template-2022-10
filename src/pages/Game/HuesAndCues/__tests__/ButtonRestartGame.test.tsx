import type { RenderResult } from '@testing-library/react'

import { fireEvent, screen, waitFor } from '@testing-library/react'
import { act } from 'react-dom/test-utils'

import { renderWithProviders } from 'helpers/test'

import { ButtonRestartGame } from '../ButtonRestartGame'
import { getMockPlayersInRoom, getMockRoom } from '../mock'
import { setRoom, setRoomPlayer } from '../services'

jest.mock('../services')

describe('<ButtonRestartGame />', () => {
  test('clickable', async () => {
    let el: RenderResult
    const room = getMockRoom()
    const players = getMockPlayersInRoom({
      p1: { isOwner: true },
    })

    await act(async () => {
      el = renderWithProviders(
        <ButtonRestartGame room={room} roomPlayers={players} currRoomPlayer={players[0]} />,
      )
    })

    const button = screen.getByTestId('button')
    expect(button).not.toHaveAttribute('disabled')
    fireEvent.click(button)

    await waitFor(() => {
      // * Update room
      expect(setRoom).toHaveBeenCalledTimes(1)
      expect(setRoom).toHaveBeenCalledWith({
        ...room,
        isStarted: false,
        hintChoice: [],
        hintSelected: '',
        isSubmitResult: false,
        totalRound: 0,
      })

      // * Update player
      expect(setRoomPlayer).toHaveBeenCalledTimes(3)
      expect(setRoomPlayer).toHaveBeenNthCalledWith(1, room.id, {
        ...players[0],
        allSelected: [],
        isHinter: true,
        isTurn: true,
        score: 0,
        totalTurn: 0,
      })
      expect(setRoomPlayer).toHaveBeenNthCalledWith(2, room.id, {
        ...players[1],
        allSelected: [],
        isHinter: false,
        isTurn: false,
        score: 0,
        totalTurn: 0,
      })
      expect(setRoomPlayer).toHaveBeenNthCalledWith(3, room.id, {
        ...players[2],
        allSelected: [],
        isHinter: false,
        isTurn: false,
        score: 0,
        totalTurn: 0,
      })

      expect(el.asFragment()).toMatchSnapshot()
    })
  })

  test('disabled, when current player is not a `owner`', async () => {
    const room = getMockRoom({ isStarted: true })
    const players = getMockPlayersInRoom({
      p1: { isOwner: true },
    })

    renderWithProviders(
      <ButtonRestartGame room={room} roomPlayers={players} currRoomPlayer={players[1]} />,
    )

    const button = screen.getByTestId('button')

    expect(button).toHaveAttribute('disabled')
  })
})
