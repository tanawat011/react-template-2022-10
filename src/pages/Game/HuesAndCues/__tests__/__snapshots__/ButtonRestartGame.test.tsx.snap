// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonRestartGame /> clickable 1`] = `
<DocumentFragment>
  <button
    data-testid="button"
  >
    Restart Game
  </button>
</DocumentFragment>
`;
